File: YacBoard/YacBoard.go
package YacBoard

import (
	"strconv"
	"unsafe"
)

const (
	Width  = 8
	Height = 8
)

type YacBoard struct {
	Fields [Width * Height]Piece

	HalfmoveClock int
	MoveNumber    int
	WhiteKingPos  int
	BlackKingPos  int
	EnPassantPos  int

	WhiteMove               bool
	WhiteCanCastleKingside  bool
	WhiteCanCastleQueenside bool
	BlackCanCastleKingside  bool
	BlackCanCastleQueenside bool
}

type byte64 struct {
	u0 uint64
	u1 uint64
	u2 uint64
	u3 uint64
	u4 uint64
	u5 uint64
	u6 uint64
	u7 uint64
}

func (board *YacBoard) Clear() {
	if strconv.IntSize == 64 && len(board.Fields) == 64 {
		ptr := (*byte64)(unsafe.Pointer(&board.Fields))
		ptr.u0 = 0
		ptr.u1 = 0
		ptr.u2 = 0
		ptr.u3 = 0
		ptr.u4 = 0
		ptr.u5 = 0
		ptr.u6 = 0
		ptr.u7 = 0
	} else {
		for i := 0; i < len(board.Fields); i++ {
			board.Fields[i] = 0
		}
	}

	board.HalfmoveClock = 0
	board.MoveNumber = 1
	board.WhiteKingPos = -1
	board.BlackKingPos = -1
	board.EnPassantPos = -1

	board.WhiteMove = true
	board.WhiteCanCastleKingside = false
	board.WhiteCanCastleQueenside = false
	board.BlackCanCastleKingside = false
	board.BlackCanCastleQueenside = false
}

func (board *YacBoard) SetField(pos int, piece Piece) {
	if uint(pos) >= Width*Height {
		panic("argument out of range")
		return
	}
	board.Fields[pos] = piece

	if piece&King == King {
		if piece == WhiteKing {
			board.WhiteKingPos = pos
		} else {
			board.BlackKingPos = pos
		}
	}
}

func (board *YacBoard) GetField(pos int) Piece {
	if uint(pos) >= Width*Height {
		return Blocked
	}
	return board.Fields[pos]
}

func (board *YacBoard) GetFieldXY(x, y int) Piece {
	if uint(x) >= Width || uint(y) >= Height {
		return Blocked
	}
	return board.GetField(x + y*Width)
}

func (board *YacBoard) GetBoardInfo() BoardInfo {
	result := BoardInfo(uint8(int8(board.EnPassantPos))) | BoardInfo(uint(board.HalfmoveClock)<<16)

	if board.WhiteCanCastleKingside {
		result |= WhiteCanCastleKingside
	}
	if board.WhiteCanCastleQueenside {
		result |= WhiteCanCastleQueenside
	}
	if board.BlackCanCastleKingside {
		result |= BlackCanCastleKingside
	}
	if board.BlackCanCastleQueenside {
		result |= BlackCanCastleQueenside
	}

	return result
}

func (board *YacBoard) SetBoardInfo(boardInfo BoardInfo) {
	board.EnPassantPos = int(int8(uint8(boardInfo & EnPassantMask)))
	board.WhiteCanCastleKingside = (boardInfo & WhiteCanCastleKingside) != BoardInfoNone
	board.WhiteCanCastleQueenside = (boardInfo & WhiteCanCastleQueenside) != BoardInfoNone
	board.BlackCanCastleKingside = (boardInfo & BlackCanCastleKingside) != BoardInfoNone
	board.BlackCanCastleQueenside = (boardInfo & BlackCanCastleQueenside) != BoardInfoNone
	board.HalfmoveClock = (int)(boardInfo&HalfmoveClockMask) >> 16
}

func (board YacBoard) String() string {
	result := make([]byte, 0, Width*Height+(4+1)*Height+6+FenMaxLength)
	for y := 0; y < Height; y++ {
		result = append(result, ' ', ' ', ' ', ' ')
		for x := 0; x < Width; x++ {
			result = append(result, PieceToChar(board.GetFieldXY(x, y)))
		}
		result = append(result, '\n')
	}
	result = append(result, "\nFEN: "...)
	result = append(result, board.GetFEN()...)
	return string(result)
}

File: dart/yacboard/lib/yacboard/yac_board.dart
import 'dart:typed_data';
import 'board_info.dart';
import 'fen.dart';
import 'board_size.dart';
import 'piece.dart';
import 'pos.dart';

class YacBoard {
  final Uint8List fields = Uint8List(BoardSize.fieldCount);

  int halfmoveClock = 0;
  int moveNumber = 1;
  int whiteKingPos = -1;
  int blackKingPos = -1;
  int enPassantPos = BoardInfo.EnPassantNone;

  bool whiteMove = true;
  bool whiteCanCastleKingside = false;
  bool whiteCanCastleQueenside = false;
  bool blackCanCastleKingside = false;
  bool blackCanCastleQueenside = false;

  YacBoard({String startFen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"}) {
    setFen(startFen);
  }

  void clear() {
    for (int i = 0; i < fields.length; i++) {
      fields[i] = 0;
    }
    halfmoveClock = 0;
    moveNumber = 1;
    whiteKingPos = -1;
    blackKingPos = -1;
    enPassantPos = BoardInfo.EnPassantNone;
    whiteMove = true;
    whiteCanCastleKingside = false;
    whiteCanCastleQueenside = false;
    blackCanCastleKingside = false;
    blackCanCastleQueenside = false;
  }

  void setField(int pos, int piece) {
    if (pos < 0 || pos >= BoardSize.fieldCount) return; // out of range
    fields[pos] = piece;
    if (piece & Piece.king != 0) {
      if (piece == Piece.whiteKing) {
        whiteKingPos = pos;
      } else {
        blackKingPos = pos;
      }
    }
  }

  int getField(int pos) {
    if (pos < 0 || pos >= BoardSize.fieldCount) return Piece.blocked;
    return fields[pos];
  }

  int getBoardInfo() {
    int result = enPassantPos | (halfmoveClock << 16);

    if (whiteCanCastleKingside) result |= BoardInfo.WhiteCanCastleKingside;
    if (whiteCanCastleQueenside) result |= BoardInfo.WhiteCanCastleQueenside;
    if (blackCanCastleKingside) result |= BoardInfo.BlackCanCastleKingside;
    if (blackCanCastleQueenside) result |= BoardInfo.BlackCanCastleQueenside;

    return result;
  }

  @override
  String toString() {
    List<int> result = [];
    for (int y = 0; y < BoardSize.height; y++) {
      result.addAll([32, 32, 32, 32]); // "    "
      for (int x = 0; x < BoardSize.width; x++) {
        result.addAll(Piece.toChar(getField(Pos.fromXY(x, y))).codeUnits);
      }
      result.add(10); // '\n'
    }
    result.addAll(("\nFEN: ${getFen()}").codeUnits);
    return String.fromCharCodes(result);
  }
}
